### PR TITLE
syncer(dm): don't check flush inside an event and refactor

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve? 
part of refactor

### What is changed and how it works?
- `addJob` now becomes a simple function of add job to channel, rest logic goes to `handleJob`
- as title
  - handleRowsEvent will save the table point and check flush after all jobs are added
  - remove a test case because we no longer flush inside an event

### Check List 

Tests 

 - Integration test

Code changes

Side effects

Related changes

### Release note 

```release-note
`None`.
```

### CODE REVIEW

**Review by lichunzhu — COMMENTED** [review]
Don't understand what does `don't check flush inside an event` means.

File: dm/syncer/syncer.go [hunk 1]
```diff
@@ -2252,12 +2275,34 @@ func (s *Syncer) handleRowsEvent(ev *replication.RowsEvent, ec eventContext) err
...
+
+			err2 := s.flushJobs()
+			if err2 != nil {
+				s.tctx.L().DPanic("flush checkpoint failed", zap.Error(err2))
+			}
+			failpoint.Return(nil)
+		}
+	})
+
+	s.checkShouldFlush()
```
Comment (lichunzhu) [inline dm/syncer/syncer.go:2305]: It seems that we run `s.checkShouldFlush()` twice for dmls?

File: dm/syncer/job.go [hunk 2]
```diff
@@ -73,6 +73,9 @@ func (t opType) String() string {
 	return ""
 }
 
+// job is the least unit of work for DMLWorker/DDL worker. It's generated by syncer's inner logic or binlog events. Specifically, most
```
[the remark this answers is not in the document]
Comment (lance6716) [reply dm/syncer/job.go:76]: I want to refer to https://github.com/pingcap/tiflow/blob/d84f15b221bf07b29c9949344f38b3248d7f17f6/dm/syncer/dml_worker.go#L33 and we don't have a DDL worker now

File: dm/syncer/syncer.go (hunk 1 above)
Comment (lance6716) [reply dm/syncer/syncer.go:2306]: fixed in https://github.com/pingcap/tiflow/pull/3891/commits/e549451fae295da5a31ddcbd923d27bd74c19933

**Review by lichunzhu — COMMENTED** [review]
Should the title be changed from
~~syncer(dm): don't check flush inside an event and refactor~~ to
syncer(dm): don't check flush inside a transaction and refactor

Besides, can you explain what can we benefit from check flush after a transaction and whether it will affect a huge transaction's retry time?

**Comment by lance6716** [comment]
> 

a transaction may has multiple events, and an event may has multiple row changes thus multiple jobs. Before this PR we check if global checkpoint should be flushed in `addJob`, so we may flush in the middle of sending an event.

For old behaviour because the event is not fully sent, we still can't skip it. So we still need replicate this event when retry. Now in position-based replication we have confidence to skip it.

**Review by lichunzhu — APPROVED** [review]
LGTM

**Review by glorv — APPROVED** [review]

**Comment by Ehco1996** [comment]
need fix lint `github.com/pingcap/tiflow/pkg/errorutil` redeclared in syncer_test.go

**Pull request merged**